imagebuildah/stage_executor.go

```diff
@@ -1521,6 +1586,124 @@ func (s *StageExecutor) tagExistingImage(ctx context.Context, cacheID, output st
+func (s *StageExecutor) pushCache(ctx context.Context, src, destSpec string) error {
+	logrus.Debugf("trying to push cache to dest: %+v from src:%+v", destSpec, src)
+	dest, err := alltransports.ParseImageName(destSpec)
+	// Add the docker:// transport to see if user neglected it while
+	// specifying --cache-to `<dest>`.
+	if err != nil {
+		destTransport := strings.Split(destSpec, ":")[0]
+		if t := transports.Get(destTransport); t != nil {
+			return err
+		}
+
+		if strings.Contains(destSpec, "://") {
+			return err
+		}
+
+		destSpec = "docker://" + destSpec
+		dest2, err2 := alltransports.ParseImageName(destSpec)
+		if err2 != nil {
+			return err
+		}
+		dest = dest2
+		logrus.Debugf("Assuming docker:// as the transport method for cache DESTINATION: %s", destSpec)
+	}
+	options := buildah.PushOptions{
+		Compression:         s.executor.compression,
+		SignaturePolicyPath: s.executor.signaturePolicyPath,
+		Store:               s.executor.store,
+		SystemContext:       s.executor.systemContext,
+		BlobDirectory:       s.executor.blobDirectory,
+		SignBy:              s.executor.signBy,
+		MaxRetries:          s.executor.maxPullPushRetries,
+		RetryDelay:          s.executor.retryPullPushDelay,
+	}
+	ref, digest, err := buildah.Push(ctx, src, dest, options)
+	if err != nil {
+		return fmt.Errorf("failed pushing cache to %q: %w", dest, err)
+	}
+	logrus.Debugf("successfully pushed cache to dest: %+v with ref:%+v and disgest: %v", dest, ref, digest)
+	return nil
+}
+
+// pullCache takes the image source of the cache assuming tag
+// already points to the valid cacheKey and pulls the image to
+// local storage only if it was not already present on local storage
+// or a newer version of cache was found in the upstream repo. If new
+// image was pulled function returns image id otherwise returns empty
+// string "" or error if any error was encontered while pulling the cache.
+func (s *StageExecutor) pullCache(ctx context.Context, src string) (string, error) {
+	logrus.Debugf("trying to pull cache from remote repo: %+v", src)
+	reportWriter := bytes.NewBufferString("")
+	options := buildah.PullOptions{
+		SignaturePolicyPath: s.executor.signaturePolicyPath,
+		Store:               s.executor.store,
+		SystemContext:       s.executor.systemContext,
+		BlobDirectory:       s.executor.blobDirectory,
+		MaxRetries:          s.executor.maxPullPushRetries,
+		RetryDelay:          s.executor.retryPullPushDelay,
+		AllTags:             false,
+		ReportWriter:        reportWriter,
+		PullPolicy:          define.PullIfNewer,
+	}
+
+	id, err := buildah.Pull(ctx, src, options)
+	if err != nil {
+		logrus.Debugf("failed pulling cache from repo %s: %v", src, err)
+		return "", fmt.Errorf("failed while pulling cache from %q: %w", src, err)
+	}
+	// Don't return any id, since image was already present on
+	// local so buildah will eventually figure out and use it as
+	// a valid cache source from local storage
+	if reportWriter.String() == "" {
```

Comment:
We need to confirm if actually image was pulled or not so that we can generate correct output text in the build output. Also we are using build output to run tests.

I have removed reportwriter and checking this against the localstorage now, but I agree that Pull API can be extended for a better implementation so will replace this with Pull API once I add support in Pull API for this but will do that in a different PR.